sdl_android/src/main/java/com/smartdevicelink/api/screen/TextAndGraphicManager.java

```diff
@@ -0,0 +1,893 @@
+				queuedUpdateListener = null;
+			}
+
+			if (listener != null){
+				queuedUpdateListener = listener;
+			}else{
+				hasQueuedUpdate = true;
+			}
+			return;
+		}
+
+		Show fullShow = new Show();
+		fullShow.setAlignment(textAlignment);
+		fullShow = assembleShowText(fullShow);
+		fullShow = assembleShowImages(fullShow);
+
+		inProgressListener = listener;
+
+		if (!shouldUpdatePrimaryImage() && !shouldUpdateSecondaryImage()){
+
+			Log.v(TAG, "No Images to send, only sending text");
+			inProgressUpdate = extractTextFromShow(fullShow);
+			sendShow();
+
+		}else if (isArtworkUploadedOrDoesntExist(primaryGraphic) && ( secondaryGraphic == blankArtwork || isArtworkUploadedOrDoesntExist(secondaryGraphic))){
+
+			Log.v(TAG, "Images already uploaded, sending full update");
+			// The files to be updated are already uploaded, send the full show immediately
+			inProgressUpdate = fullShow;
+			sendShow();
+		} else{
+
+			Log.v(TAG, "Images need to be uploaded, sending text and uploading images");
+			// We need to upload or queue the upload of the images
+			// start uploading images
+			inProgressUpdate = fullShow;
+			final Show thisUpdate = fullShow;
+
+			uploadImages(new CompletionListener() {
+				@Override
+				public void onComplete(boolean success) {
+					if (!success){
+						Log.e(TAG, "Error uploading image");
+						inProgressUpdate = extractTextFromShow(inProgressUpdate);
+						sendShow();
+					}
+					// Check if queued image update still matches our images (there could have been a new Show in the meantime)
+					// and send a new update if it does. Since the images will already be on the head unit, the whole show will be sent
+					if ((thisUpdate.getGraphic() != null && queuedImageUpdate.getGraphic() != null) && thisUpdate.getGraphic().equals(queuedImageUpdate.getGraphic()) ||
+							(thisUpdate.getSecondaryGraphic() != null && queuedImageUpdate.getSecondaryGraphic() != null) && thisUpdate.getSecondaryGraphic().equals(queuedImageUpdate.getSecondaryGraphic())){
+						Log.v(TAG, "Queued image update matches the images we need, sending update");
+						sendShow();
+					} else {
+						Log.v(TAG, "Queued image update does not match the images we need, skipping update");
+					}
+				}
+			});
+			queuedImageUpdate = fullShow;
+		}
+	}
+
+	private void sendShow(){
+		inProgressUpdate.setOnRPCResponseListener(new OnRPCResponseListener() {
+			@Override
+			public void onResponse(int correlationId, RPCResponse response) {
+				if (response.getSuccess()){
+					Log.v(TAG, "Show Successful");
+					updateCurrentScreenDataFromShow(inProgressUpdate);
+				}
+
+				inProgressUpdate = null;
+				if (inProgressListener != null){
+					inProgressListener.onComplete(true);
+					inProgressListener = null;
+				}
+
+				if (hasQueuedUpdate){
+					Log.v(TAG, "Queued update exists, sending another update");
+					sdl_update(queuedUpdateListener);
+					queuedUpdateListener = null;
+					hasQueuedUpdate = false;
+				}
```

Comment:
These two value writes will clear out whatever was set in the sdl_update method. I would suggest moving these into the `sdl_update` method instead after the first few checks or create a temp value for the queuedUpdateListener and moving the write logic before the `sdl_update` call. 

```
hasQueuedUpdate = false;
CompletionListener temp = queuedUpdateListener;
queuedUpdateListener = null;
sdl_update(temp);
```